File: Data Structures & Algorithms/subsets-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        unordered_map<int, int> freq;
        for(auto& num: nums) {
            freq[num]++;
        }

        vector<vector<int>> res;
        res.push_back(vector<int>());

        for(auto& p: freq) {
            vector<vector<int>> combinations;
            for(int i = 1; i <= p.second; ++i) {
                combinations.push_back(vector<int>(i, p.first));
            }

            auto curr_res = res;
            for(auto& cr: curr_res) {
                for(auto& c: combinations) {
                    vector<int> new_combination;
                    new_combination.insert(new_combination.end(), cr.begin(), cr.end());
                    new_combination.insert(new_combination.end(), c.begin(), c.end());
                    res.push_back(new_combination);
                }
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        l, r = [], [1] * len(nums)
        prod = 1
        for num in nums:
            prod = prod * num
            l.append(prod)
        
        prod = 1
        for i in range(len(nums) - 1, -1, -1):
            prod = prod * nums[i]
            r[i] = prod
        
        res = [1] * len(nums)
        for i in range(1, len(nums) - 1):
            res[i] = l[i - 1] * r[i + 1]

        res[0], res[-1] = r[1], l[-2]
        return res